sycl/source/detail/graph_impl.hpp

```diff
@@ -28,29 +28,15 @@ namespace oneapi {
 namespace experimental {
 namespace detail {
 
-class wrapper {
-  using T = std::function<void(sycl::handler &)>;
-  T MFunc;
-  std::vector<sycl::event> MDeps;
-
-public:
-  wrapper(T Func, const std::vector<sycl::event> &Deps)
-      : MFunc(Func), MDeps(Deps){};
-
-  void operator()(sycl::handler &CGH) {
-    CGH.depends_on(MDeps);
-    std::invoke(MFunc, CGH);
-  }
-};
-
 struct node_impl {
-  bool MScheduled;
-
   std::shared_ptr<graph_impl> MGraph;
-  sycl::event MEvent;
+  /// ID representing this node in the graph
+  /// TODO this should be attached to an executable graph, rather than
+  /// a modifiable graph
+  pi_ext_sync_point MPiSyncPoint;
 
   std::vector<std::shared_ptr<node_impl>> MSuccessors;
-  std::vector<std::shared_ptr<node_impl>> MPredecessors;
+  std::vector<std::weak_ptr<node_impl>> MPredecessors;
```

Comment:
Previously copies of the node were added to the predecessor list in a new `shared_ptr`, rather than using the same shared_ptr used everywhere else. I removed this to prevent problems with the node copies getting out of sync, however that would have created a circular dependency between nodes (NodeA has a shared_ptr to NodeB as a successor and NodeB has a shared_ptr to NodeA as a predecessor). So switching to weak_ptrs solved that.

Ideally probably the `shared_ptr` for each node should live in a master list and both successors and predecessors should be `weak_ptr`s but that's not really relevant for functionality here.

Reply:
Thanks for the explanation 👍 
Realize I should've said a *code* comment in my feedback, since if I have this question other readers probably will too. Can be briefer like:
```cpp
// Prevents circular dependency between nodes being each others predecessor and sucessor
```